doc/upgrade-guide/README.md

```diff
@@ -1,10 +1,135 @@
 # Upgrade Guide
 
-The purpose of this guide is to detail the changes made by the successive versions of the DataStax C# Driver that 
-are relevant to for an upgrade from prior versions.
+The purpose of this guide is to detail the changes made by the successive versions of the DataStax C# Driver that are relevant to for an upgrade from prior versions.
 
 If you have any question or comment, you can [post it on the mailing list][mailing-list].
 
+## Our policy regarding API changes and release versioning
+
+The driver versions follow semantic versioning.
+
+### Major releases
+
+Example: `3.0.0`
+
+Regarding `major` releases, any public component of the driver might be changed or removed although we will always avoid introducing significant changes that would make it signifcantly harder for an application to upgrade to the new `major` version.
+
+### Minor releases
+
+Example: `3.10.0`
+
+For `minor` releases, it's a little more complicated because we don't want to be forced into bumping the `major` version every time we want to add a new client facing feature to the driver. Because of this we group the public interfaces of the driver in two categories. Here we refer to them as _mockable_ and _implementable_ but these names are here just to make it easier to explain them in this section.
+
+#### _Implementable_ Interfaces
+
+These interfaces exist to allow client applications to provide customized behavior that will be executed by the driver. For these interfaces, the driver provides ways for the applications to plug them in. Some examples: policies such as `ILoadBalancingPolicy` can be plugged in through builder methods like `Builder.WithLoadBalancingPolicy`, `ITimestampGenerator` can be plugged in through `Builder.WithTimestampGenerator`.
+
+In `minor` releases, these interfaces will not contain any change to existing methods. Additionally, _Implementable_ interfaces will never receive new methods.
+
+The reason why we commit to never add new methods to these interfaces in `minor` releases is due to the fact that these interfaces are meant to be implemented by client applications while the remaining interfaces are there mostly to make it easier to mock the driver.
+
+#### _Mockable_ Interfaces
+
+For the remaining policies, i.e., those who can not be plugged in to the driver, they exist to allow client applications to mock the driver in their test suites. These interfaces are usually the main entry points of the driver's public API with which client applications interact to execute requests. Some examples: `ISession`, `ICluster`, `IMapper`, `ICqlQueryAsyncClient`, `ICqlWriteAsyncClient`, `ICqlQueryClient`, `ICqlWriteClient`.
```

Comment:
> allow client applications to mock the driver in their test suites

mock and DI:

"allow client applications to mock the driver in their test suites and inject those dependencies in the application when needed"

Reply:
Oh, forgot about DI, nice 👍